source/fuzz/transformation_inline_function.cpp

```diff
@@ -0,0 +1,176 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_inline_function.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationInlineFunction::TransformationInlineFunction(
+    const spvtools::fuzz::protobufs::TransformationInlineFunction& message)
+    : message_(message) {}
+
+TransformationInlineFunction::TransformationInlineFunction(
+    const std::map<uint32_t, uint32_t>& result_id_map,
+    uint32_t function_call_id) {
+  *message_.mutable_result_id_map() =
+      google::protobuf::Map<google::protobuf::uint32, google::protobuf::uint32>(
+          result_id_map.begin(), result_id_map.end());
+  message_.set_function_call_id(function_call_id);
+}
+
+bool TransformationInlineFunction::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // The values in the |message_.result_id_map| must be all fresh and all
+  // distinct.
+  std::set<uint32_t> ids_used_by_this_transformation;
+  for (auto& pair : message_.result_id_map()) {
+    if (!CheckIdIsFreshAndNotUsedByThisTransformation(
+            pair.second, ir_context, &ids_used_by_this_transformation)) {
+      return false;
+    }
+  }
+
+  // |function_call_instruction| must be defined and must be an OpFunctionCall
+  // instruction.
+  auto function_call_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.function_call_id());
+  if (function_call_instruction == nullptr ||
+      function_call_instruction->opcode() != SpvOpFunctionCall) {
+    return false;
+  }
+
+  // The called function must not have an early return.
+  auto called_function = fuzzerutil::FindFunction(
+      ir_context, function_call_instruction->GetSingleWordInOperand(0));
+  if (called_function->HasEarlyReturn()) {
+    return false;
+  }
+
+  // |message_.result_id_map| must have an entry for every result id in the
+  // called function.
+  for (auto& block : *called_function) {
+    // Since the entry block label will not be inlined, only the remaining
+    // labels must have a corresponding value in the map.
+    if (&block != &*called_function->entry() &&
+        !message_.result_id_map().count(block.GetLabel()->result_id())) {
+      return false;
+    }
+    for (auto& instruction : block) {
+      if (instruction.HasResultId() &&
+          instruction.result_id() != called_function->GetReturnValueId() &&
```

Comment:
I would store the result of `called_function->GetReturnValueId()` in a temporary, computed before the outer loop, to avoid computing it again and again for every instruction that has a result id.

Reply:
Done.